src/debug.cc

```diff
@@ -71,19 +98,359 @@ class DebugFile
+
+private:
+    bool opened = false; ///< whether openlog() was called
+};
+
+/// Manages private module state that must be available during program startup
+/// and (especially) termination. Any non-trivial state objects must be
+/// encapsulated here because debugs() may be called before dynamic
+/// initialization or after the destruction of static objects in debug.cc.
+class DebugModule
+{
+public:
+    DebugModule();
+
+    // we provide debugging services for the entire duration of the program
+    ~DebugModule() = delete;
+
+    /// \copydoc Debug::PrepareToDie()
+    void prepareToDie();
+
+    /// Log the given debugs() message to appropriate channel(s) (eventually).
+    /// Assumes the message has passed the global section/level filter.
+    void log(const DebugMessageHeader &, const CompiledDebugMessageBody &);
+
+    /// Start using an open cache_log file as the primary debugs() destination.
+    /// Stop using stderr as a cache_log replacement (if we were doing that).
+    void useCacheLog();
+
+    /// Start using stderr as the primary debugs() destination.
+    /// Stop waiting for an open cache_log file (if we were doing that).
+    void banCacheLogUse();
+
+public:
+    CacheLogChannel cacheLogChannel;
+    StderrChannel stderrChannel;
+    SyslogChannel syslogChannel;
+};
+
+/// Maintains the number of code paths on the current call stack that need
+/// protection from new debugs() calls. Squid cannot _block_ re-entrant debugs()
+/// calls, but the high-level debugs() handling code queues re-entrant logging
+/// attempts when Busy() instead of letting them through to sensitive code.
+class LoggingSectionGuard
+{
+public:
+    LoggingSectionGuard();
+    ~LoggingSectionGuard();
+
+    /// whether new debugs() messages must be queued
+    static bool Busy() { return LoggingConcurrencyLevel; }
+
+private:
+    /// the current number of protected callers
+    static size_t LoggingConcurrencyLevel;
+};
+
+size_t LoggingSectionGuard::LoggingConcurrencyLevel = 0;
+
+/// debugs() messages postponed due to LoggingSectionGuard::Busy(). This is the
+/// head of the invasive Context::upper FIFO list of such messages.
+static Debug::Context *WaitingForIdle = nullptr;
+
+/// cache_log file
 /// safe during static initialization, even if it has not been constructed yet
+/// safe during program termination, even if it has been destructed already
 static DebugFile TheLog;
 
 FILE *
 DebugStream() {
-    return TheLog.file();
+    return TheLog.file() ? TheLog.file() : stderr;
+}
+
+/// used for the side effect: fills Debug::Levels with the given level
+static void
+ResetSections(const int level)
+{
+    for (auto &sectionLevel: Debug::Levels)
+        sectionLevel = level;
+}
+
+/* LoggingSectionGuard */
+
+LoggingSectionGuard::LoggingSectionGuard()
+{
+    ++LoggingConcurrencyLevel;
+}
+
+LoggingSectionGuard::~LoggingSectionGuard()
+{
+    if (--LoggingConcurrencyLevel == 0)
+        Debug::LogWaitingForIdle();
+}
```

Comment:
Please inline these to the local class definition.

Reply:
Why? I think they would distract from the code comprehension if inlined, especially since one of the paired methods is a multi-line statement with non-trivial logic.